### PR TITLE
HPCC-20043 Roxie tracing of excessive seeks does not handle variable records well

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Manual testing.

### CODE REVIEW

File: system/jhtree/jhtree.cpp [hunk 1]
```diff
@@ -2088,7 +2088,7 @@ void IndexRowFilter::finish(size32_t _keyedSize)
 
 void IndexRowFilter::describe(StringBuffer &out) const
 {
-    for (unsigned idx=0; idx < lastRealSeg(); idx++)
+    for (unsigned idx=0; idx <= lastRealSeg(); idx++)
```
Comment (richardkchapman) [inline system/jhtree/jhtree.cpp]: This is a separate bug spotted during testing

**Comment by richardkchapman** [comment]
@ghalliday Please review. We could consider a fix in 6.4.x but it would probably be a different PR - code is completely different around here.

File: system/jhtree/jhtree.cpp [hunk 2]
```diff
@@ -112,7 +112,7 @@ SegMonitorList::SegMonitorList(const SegMonitorList &from, const char *fixedVals
 
 void SegMonitorList::describe(StringBuffer &out) const
 {
-    for (unsigned idx=0; idx < lastRealSeg(); idx++)
+    for (unsigned idx=0; idx <= lastRealSeg(); idx++)
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: Won't this crash if there are no segment monitors? (I think lastRealSeg still returns 0 and the item call will fail.)

File: system/jhtree/jhtree.cpp (hunk 1 above)
Comment (ghalliday) [reply system/jhtree/jhtree.cpp]: similar problem if there are no filters.

**Comment by ghalliday** [comment]
@richardkchapman see comments on one of the changes

File: system/jhtree/jhtree.cpp (hunk 2 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: Not sure it is ever called in a case where no filters, but I will look at making more robust

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

**Pull request merged**